### ISSUE TITLE
Question: How to start VirtualBox on Windows?

### ISSUE THREAD

**Issue by nccurry** [code-ready/crc#751]
Note: While CRC provides a driver for VirtualBox, this is an unsupported and mostly untested driver. We have received positive results from users during the alpha and beta testing period. We have included the driver for those who are willing to tinker when a native hypervisor is unavailable, such as due to company restrictions and/or policies. But be aware, from our own tests we conclude that VirtualBox overall consumes more memory and is slower in use than the native options we provide.  — **CRC Team**

---

### General information

  * OS: Linux / macOS / Windows
    * Windows
  * Hypervisor: KVM / Hyper-V / VirtualBox / hyperkit
    * VirtualBox
  * Did you run `crc setup` before starting it (Yes/No)?
    * Yes

## CRC version
```bash
version: 1.0.0-rc.0+34371d3
OpenShift version: 4.2.0-0.nightly-2019-09-26-192831 (embedded in binary)
```
  
## CRC status
```bash
ERRO Error occurred: Unable to connect to the server: dial tcp: lookup api.crc.testing: no such host
 - exit status 1
```

## CRC config
```bash
```

## Host Operating System
```bash
Host Name:                 RHT460S
OS Name:                   Microsoft Windows 10 Pro
OS Version:                10.0.18362 N/A Build 18362
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          John Doe
Registered Organization:
Product ID:                00330-80000-00000-AA835
Original Install Date:     8/23/2019, 11:44:55 PM
System Boot Time:          10/15/2019, 8:49:58 AM
System Manufacturer:       LENOVO
System Model:              20FAS1EF0M
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: Intel64 Family 6 Model 78 Stepping 3 GenuineIntel ~2607 Mhz
BIOS Version:              LENOVO N1CET77W (1.45 ), 8/6/2019
Windows Directory:         C:\Windows
System Directory:          C:\Windows\system32
Boot Device:               \Device\HarddiskVolume1
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-06:00) Central Time (US & Canada)
Total Physical Memory:     19,880 MB
Available Physical Memory: 5,692 MB
Virtual Memory: Max Size:  22,824 MB
Virtual Memory: Available: 6,654 MB
Virtual Memory: In Use:    16,170 MB
Page File Location(s):     C:\pagefile.sys
Domain:                    WORKGROUP
Logon Server:              \\RHT460S
Hotfix(s):                 10 Hotfix(s) Installed.
                           [01]: KB4515871
                           [02]: KB4503308
                           [03]: KB4506472
                           [04]: KB4508433
                           [05]: KB4509096
                           [06]: KB4515383
                           [07]: KB4515530
                           [08]: KB4516115
                           [09]: KB4521863
                           [10]: KB4517389
Network Card(s):           8 NIC(s) Installed.
                           [01]: Intel(R) Ethernet Connection I219-LM
                                 Connection Name: Ethernet
                                 DHCP Enabled:    Yes
                                 DHCP Server:     192.168.1.254
                                 IP address(es)
                                 [01]: 192.168.1.118
                                 [02]: fe80::9425:3fee:3458:c62e
                           [02]: Intel(R) Dual Band Wireless-AC 8260
                                 Connection Name: Wi-Fi
                                 Status:          Media disconnected
                           [03]: Bluetooth Device (Personal Area Network)
                                 Connection Name: Bluetooth Network Connection
                                 Status:          Media disconnected
                           [04]: VirtualBox Host-Only Ethernet Adapter
                                 Connection Name: VirtualBox Host-Only Network
                                 DHCP Enabled:    No
                                 IP address(es)
                                 [01]: 192.168.130.1
                                 [02]: fe80::885:d9e6:70a1:b55
                           [05]: TAP-Windows Adapter V9
                                 Connection Name: Ethernet 2
                                 Status:          Media disconnected
                           [06]: PANGP Virtual Ethernet Adapter
                                 Connection Name: Ethernet 3
                                 Status:          Media disconnected
                           [07]: VMware Virtual Ethernet Adapter for VMnet1
                                 Connection Name: VMware Network Adapter VMnet1
                                 DHCP Enabled:    Yes
                                 DHCP Server:     192.168.245.254
                                 IP address(es)
                                 [01]: 192.168.245.1
                                 [02]: fe80::ed76:1485:bf17:fcb4
                           [08]: VMware Virtual Ethernet Adapter for VMnet8
                                 Connection Name: VMware Network Adapter VMnet8
                                 DHCP Enabled:    Yes
                                 DHCP Server:     192.168.101.254
                                 IP address(es)
                                 [01]: 192.168.101.1
                                 [02]: fe80::2157:118f:b5ab:7f47
Hyper-V Requirements:      VM Monitor Mode Extensions: Yes
                           Virtualization Enabled In Firmware: Yes
                           Second Level Address Translation: Yes
                           Data Execution Prevention Available: Yes
```

### Steps to reproduce

  1. ./crc setup --vm-driver virtualbox
  2. .\crc.exe start --vm-driver virtualbox --bundle .\crc_virtualbox_4.2.0-0.nightly-2019-09-26-192831.crcbundle --pull-secret-file .\pull-secret.txt

### Expected
Working crc

### Actual
Non-working crc

### Logs

```
INFO Starting OpenShift cluster ... [waiting 3m]
INFO
INFO To access the cluster, first set up your environment by following 'crc oc-env' instructions
INFO Then you can access it by running 'oc login -u developer -p developer https://api.crc.testing:6443'
INFO To login as an admin, username is 'kubeadmin' and password is F44En-Xau6V-jQuyb-yuMXB
INFO
INFO You can now run 'crc console' and use these credentials to access the OpenShift web console
ERRO Error occurred: Error approving the node csr Not able to get csr names (exit status 1 : Unable to connect to the server: dial tcp: lookup api.crc.testing: no such host
)
Making call to close driver server
(crc) Calling .Close
Successfully made call to close driver server
Making call to close connection to plugin binary
Making call to close driver server
(crc) Calling .Close
(crc) DBG | Closing plugin on server side
Successfully made call to close driver server
Making call to close connection to plugin binary
```

**Comment by praveenkumar** [comment]
@nccurry for virtualbox on windows, we are not setting up the dns server so in that case you have to set it up yourself https://code-ready.github.io/crc/#_dns_configuration_details should satisfy these domains. Is there any reason you are not using the hyperV instead virtualbox ?

**Comment by nccurry** [comment]
I do not see anything in there covering Windows 10.

I'm assuming I need to add DNS entries for the API and infrastructure routes to ```C:\Windows\System32\drivers\etc\hosts```?

> Is there any reason you are not using the hyperV instead virtualbox ?

Hyper-V has an abhorrent user experience compared to VirtualBox or VMware Workstation, so I prefer to use those if possible.

When Hyper-V is enabled it causes VirtualBox to default to using it as a backend, which has significantly slower performance, so I keep it disabled so as to not interfere with other VMs I use in VirtualBox.

**Comment by gbraad** [comment]
Note: in case of another hypervisor, VirtualBox will default to use full system virtualization (using the emulation backend) as it does not have access to hypercalls. be aware, that when WSL2 is enabled in future (or from Insider) you are unable to use VirtualBox in that case.

you would need to add address records for the `api` and `console` to allow access. They aren't necessary for use of the cluster, as it relies on its internal DNS settings, but they are needed for remote access. You can also try to use `Set-DNSClientServerAddress` on one of the virtual network adapters of VirtualBox to allow resolving.

in future releases we might automate this in a more generic way (as a fallback option) when we can't set the nameserver on an interface.

**Comment by praveenkumar** [comment]
> I'm assuming I need to add DNS entries pointing to the crc VM for the API and infrastructure routes to `C:\Windows\System32\drivers\etc\hosts`?

@nccurry yes, you need the to put entry in the `hosts` file in this case and it should have atleast api and oauth end points. Can you try something following and see if that works?

- Get the ip of crc vm `crc ip`
- Add the entry to `hosts`.
```
<crc ip> api.crc.testing oauth-openshift.apps-crc.testing
```
- Try to do login using `oc login -u kubeadmin -p <password_kubeadmin>

**Comment by cfergeau** [comment]
I believe we can retitle this bug to "Missing documentation about expected DNS setup on Windows"?

**Comment by nccurry** [comment]
Trying ```./crc.exe ip --log-level=debug``` results in:

```(crc) DBG | ip addr show
(crc) DBG | Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [publickey none], no supported methods remain
(crc) DBG | Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
(crc) DBG | Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
(crc) DBG | Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
(crc) DBG | Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

Looking at the VM in VirtualBox I can see two interfaces ```enp0s3``` and ```enp0s8``` and their ip addresses.
![image](https://user-images.githubusercontent.com/3799024/67298884-80467280-f4b1-11e9-84ba-7086b9045599.png)

Trying to run oc against those interfaces directly (or modifying ```C:\Windows\System32\drivers\etc\hosts``` directly as suggested above) results in:

```
PS C:\Users\User\Documents\crc> .\oc.exe login --insecure-skip-tls-verify=true -u 'kubeadmin' -p 'F44En-Xau6V-jQuyb-yuMXB' https://10.0.2.15:8443
error: dial tcp 10.0.2.15:8443: connectex: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. - verify you have provided the correct host and port and that the server is currently running.

PS C:\Users\User\Documents\crc> .\oc.exe login --insecure-skip-tls-verify=true -u 'kubeadmin' -p 'F44En-Xau6V-jQuyb-yuMXB' https://192.168.130.106:8443
error: dial tcp 192.168.130.106:8443: connectex: No connection could be made because the target machine actively refused it. - verify you have provided the correct host and port and that the server is currently running.
```

**Comment by gbraad** [comment]
VirtualBox is an unsupported target so your mileage may vary. If you want a tested, well working out of the box experience, consider using Hyper-V.

**Comment by gbraad** [comment]
Note: likely due to `(crc) DBG | Error dialing TCP: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain` the start might have already failed along the way. Please provide full log output if you need assistance.

**Comment by gbraad** [comment]
@cfergeau the OP keeps adding errors related to running VB, so I'd rather create a new issue for this and use this to track/reply to the errors he encounters

**Comment by praveenkumar** [comment]
> PS C:\Users\User\Documents\crc> .\oc.exe login --insecure-skip-tls-verify=true -u 'kubeadmin' -p 'F44En-Xau6V-jQuyb-yuMXB' https://192.168.130.106:8443

@nccurry Port should be `6443` not `8443`, please try using below after you add `192.168.130.106` in your `/etc/hosts`

```
> .\oc.exe login --insecure-skip-tls-verify=true -u 'kubeadmin' -p 'F44En-Xau6V-jQuyb-yuMXB' api.crc.testing:6443
```

**Comment by nccurry** [comment]
Ah good call.

So going into the virtualbox vm console and getting the ip address, putting the following in```C:\Windows\System32\drivers\etc\hosts```, and using the correct port seems to make everything functional.,

```
<ip> api.crc.testing 
<ip> api-int.crc.testing 
<ip> oauth-openshift.apps-crc.testing 
<ip> console-openshift-console.apps-crc.testing 
<ip> downloads-openshift-console.apps-crc.testing
<ip> default-route-openshift-image-registry.apps-crc.testing
<ip> alertmanager-main-openshift-monitoring.apps-crc.testing
<ip> grafana-openshift-monitoring.apps-crc.testing
<ip> prometheus-k8s-openshift-monitoring.apps-crc.testing
```

```PS C:\Users\User\Documents\crc> .\oc.exe login --insecure-skip-tls-verify=true -u 'kubeadmin' -p 'F44En-Xau6V-jQuyb-yuMXB' https://192.168.130.107:6443```

**Comment by gbraad** [comment]
@praveenkumar it would be enough to add:

```
<ip> api.crc.testing 
<ip> api-int.crc.testing 
<ip> oauth-openshift.apps-crc.testing 
<ip> console-openshift-console.apps-crc.testing 
```

right? What about macOS?

**Comment by praveenkumar** [comment]
> @praveenkumar it would be enough to add:
> 
> ```
> <ip> api.crc.testing 
> <ip> api-int.crc.testing 
> <ip> oauth-openshift.apps-crc.testing 
> <ip> console-openshift-console.apps-crc.testing 
> ```
> 
> right? What about macOS?

We don't really need to add `api-int.crc.testing` since this is for internal communication. Also `console-openshift-console.apps-crc.testing` we resolve using the mac resolver.

**Comment by gbraad** [comment]
```
<ip> api.crc.testing 
<ip> oauth-openshift.apps-crc.testing 
<ip> console-openshift-console.apps-crc.testing 
```

So that would be the essential list, before any application is deployed when using `hosts` as an alternative source for the DNS.

Note: while console is handles by the resolver, so is api and oauth. we add these entries for fallback purposes and in case of vbox.

**Comment by gbraad** [comment]
VirtualBox support has been deprecated, See #838

**Comment by rob-4x4** [comment]
> VirtualBox support has been deprecated, See #838

Very poor ... Hyper-V is horrible to use compared to Virtual box or vmware ...
